specs/consensus.md

```diff
@@ -0,0 +1,22 @@
+Consensus Rules
+===
+
+- [Consensus Rules](#consensus-rules)
+- [System Parameters](#system-parameters)
+- [Leader Selection](#leader-selection)
+- [Fork Choice](#fork-choice)
+- [Block Validity](#block-validity)
+
+# System Parameters
+
+| name                         | type     | value | description                                                      |
+| ---------------------------- | -------- | ----- | ---------------------------------------------------------------- |
+| `SHARE_SIZE`                 | `uint64` | `256` | Size of transaction and message shares, in bytes.                |
+| `SHARE_RESERVED_BYTES`       | `uint64` | `1`   | Bytes reserved at the beginning of each share.                   |
+| `INTERMEDIATE_ROOT_INTERVAL` | `uint64` | `1`   | Maximum number of transactions between intermediate state roots. |
+
+# Leader Selection
+
```

Comment:
Here is a description of the proposer selection in tendermint: https://docs.tendermint.com/master/spec/reactors/consensus/proposer-selection.html

Reply:
Note that from looking at tendermint as a more general library, it would be nicer if users could more easily change the underlying proposer election mechanism. I thought there was an issue about this (and that the proposer election algo should be defined as a pure function of height, round, and last proposer, but couldn't find it). In any case, this could look sth along this lines: https://github.com/celo-org/celo-blockchain/blob/ffa8d5ea42d26cd0636eda197f37b101598c0de4/consensus/istanbul/validator.go#L125:6

This would also be a refactoring that could be contributed back upstream.

Reply:
Not sure yet if we want to go the round-robin or randomized (Tenderand?) route. I don't inherently see any major issues with round-robin since the leader is predictable even if randomized.

The consensus protocol used by apps to internally select block proposers for apps (which are then passed to LazyLedger block proposers) is orthogonal, but we should definitely keep it in mind in case there are common design decisions that can be made.